### PR TITLE
[state dict] Change _load_model_state_dict to enable cpu_offload, accept 2 device type and optimize memory

### PR DESCRIPTION
**Pull request by mori360**
For destributed state dict api [migration](https://github.com/pytorch/torchtune/pull/2138), make the changes here:
1. `load_from_full_model_state_dict` at TorchTune calls `set_model_state_dict` with the options on whether to have cpu_offload. Add cpu_offload at _load_model_state_dict to process to cpu if config is True
2. Change the device check as lora_finetune might hace 2 device types, accept that to be valid.
3. Some changes to optimize the memory performance:
3.1 use `.detach().clone()` instead of view directly
3.2 if local_state is not meta, copy `full_tensor[slices]` to `ret.to_local()`
4. add relative unit tests

Memory performance calling from TorchTune with llama2/7B_full:
1. cpu_offload = True
Screenshot 2024-12-18 at 1 36 47 PM

2. cpu_offload = False
Screenshot 2024-12-18 at 1 36 52 PM

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

File: torch/distributed/_state_dict_utils.py [hunk 1]
```diff
@@ -560,16 +560,21 @@ def _distribute_tensors(
...
+            ret = DTensor.from_local(
+                local_tensor,
+                local_state.device_mesh,
+                local_state.placements,
+                shape=local_state.shape,
+                stride=local_state.stride(),
+            )
+        else:
+            ret = local_state
+            ret.to_local().copy_(full_tensor[slices])
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:576]: We need some comments to explain the motivation of this block.

**Review by fegin — COMMENTED** [review]
I think we should retitle the PR. It's not just cpu_offload.

File: torch/distributed/_state_dict_utils.py [hunk 2]
```diff
@@ -560,16 +560,21 @@ def _distribute_tensors(
...
-        # one of the case that is not yet supported is when placements = (Shard(0), _StridedShard(0, sf=2)).
-        local_state_dict[key] = DTensor.from_local(
-            local_tensor,
-            local_state.device_mesh,
-            local_state.placements,
-            shape=local_state.shape,
-            stride=local_state.stride(),
-        )
+        if local_state.is_meta:
+            local_tensor = full_tensor[slices].detach().clone()
```
Comment (fegin) [inline torch/distributed/_state_dict_utils.py:564]: Also need a comment why we need clone.

File: torch/distributed/checkpoint/state_dict.py [hunk 3]
```diff
@@ -555,19 +555,34 @@ def _load_model_state_dict(
...
-        assert device is not None
-        if device == torch.device("meta"):
-            device = dist.distributed_c10d._get_pg_default_device()
-            assign = True
+                devices.add(value.device)
+        if len(devices) == 1:
+            device = devices.pop()
+            if device == torch.device("meta"):
+                device = dist.distributed_c10d._get_pg_default_device()
+                assign = True
```
Comment (fegin) [inline torch/distributed/checkpoint/state_dict.py:566]: Do you think that if we always do `assign=True`, is it better?

File: torch/distributed/checkpoint/state_dict.py [hunk 4]
```diff
@@ -555,19 +555,34 @@ def _load_model_state_dict(
...
+        # In lora state_dict, ther could be multiple devices, with meta device inside.
```
Comment (fegin) [inline torch/distributed/checkpoint/state_dict.py:567]: nits: ther->there

File: torch/distributed/checkpoint/state_dict.py [hunk 5]
```diff
@@ -555,19 +555,34 @@ def _load_model_state_dict(
     assign = False
     if info.broadcast_from_rank0 or info.full_state_dict:
         device = None
```
Comment (fegin) [inline torch/distributed/checkpoint/state_dict.py:557]: You don't need to have this `device` variable anymore. 
You can just change the line
```
device=device,
```
to
```
device=devices.pop(),
```

Comment (mori360) [reply torch/distributed/checkpoint/state_dict.py]: Change the logic here to remove deivce and just use devices.pop() for the device input

File: test/distributed/checkpoint/test_state_dict.py [hunk 6]
```diff
@@ -875,6 +875,55 @@ def test_setting_meta_device_model_broadcasting(self) -> None:
             )
             self.assertEqual(cpu_model_value, meta_model_value)
 
+    @with_comms
+    @skip_if_lt_x_gpu(1)
+    def test_memory_between_view_and_clone(self) -> None:
```
[the remark this answers is not in the document]
Comment (mori360) [reply test/distributed/checkpoint/test_state_dict.py]: Add memory check at `test_setting_meta_device_model_broadcasting` after calling `set_model_state_dict`
Test pass right now by the change if 'clone', but would fail if without the change in this PR

File: torch/distributed/checkpoint/state_dict.py (hunk 3 above)
Comment (mori360) [reply torch/distributed/checkpoint/state_dict.py]: test_shared_weight would fail id we set assign to True by default.
We can change assign and it's related tests in the future PR.

**Review by fegin — APPROVED** [review]
LGTM, thanks for the fix.
